Switch to Scalafmt

patriknw opened the pull request:
* and replace unicode arrows

patriknw committed:
Switch to Scalafmt in build

* AvoidInfix in scalafmt.conf

* align.tokens in scalafmt.conf

* scalafmt in MultiJvm

* sbt-scalafmt 2.0.0-RC5

* scalafmt in all* command alias

* scalafmtAll in CONTRIBUTING.md

patriknw committed:
manual adjustments before scalafmt
* fix formatting error

* reversePrepend instead of reverse_:::
  * because it is causing trouble for AvoidInfix formatting

* prepare for AvoidInfix

* fix try-catch

patriknw committed:
replace unicode arrows

* ⇒, →, ←
* because we don't want to show them in documentation snippets and
  then it's complicated to avoid that when snippets are
  located in src/test/scala in individual modules
* dont replace object `→` in FSM.scala and PersistentFSM.scala

patriknw commented:
One part is missing here, it should also scalafmtOnCompile for multi-jvm sources. Probably some kind of `dependsOn`.

patriknw commented:
The performance seems ok since it does a good job at format incrementally. 

a full `scalafmt` takes ~50 seconds

If it turns out to be too slow we can disable format on compile, which is what is recommended in docs https://scalameta.org/scalafmt/docs/installation.html#format-on-compile.

Format on save in IntelliJ is probably good to enable, anyway: https://scalameta.org/scalafmt/docs/installation.html#format-on-save

2m reviewed (approved):
LGTM.

2m commented:
There has recently been improvements to sbt-scalafmt's formatting of code in non-standard configurations and also improved cached formatting:

https://github.com/scalameta/sbt-scalafmt/pull/14
https://github.com/scalameta/sbt-scalafmt/pull/13

Improvements not yet released, tho.

2m commented:
This should enable the formatting for `MutiJvm` configuration:

```scala
settings(
   inConfig(MultiJvm)(scalafmtConfigSettings)
   ...
)
```

helena commented on .scalafmt.conf:
```diff
@@ -0,0 +1,13 @@
+docstrings                 = JavaDoc
+indentOperator             = spray
+maxColumn                  = 120
+rewrite.rules              = [RedundantParens, SortImports]
```
You seem to prefer `AvoidInfix` as well, but we may not want to rewrite those.

helena replied:
My strong pref is for `RedundantBraces` as well but...

patriknw replied:
> RedundantBraces

I think that depends how large the statement is. If it's a real one-liner then yes, but if it's wrapping over several lines and having comments I prefer some extra `{ }` to clarify the scope.

patriknw replied:
I would like to use `AvoidInfix` but it is not nice for some things, like
```
withAttributes(traversalBuilder.attributes.and(attr))
```

but maybe the total result is better? I have pushed that as a separate commit so we can compare and decide: 04ae484

patriknw replied:
There were some more trouble with AvoidInfix that I have solved in 32199ce and 1627f14

helena replied:
`AvoidInfix` seems good now, WDYT?

patriknw replied:
`rewrite.neverInfix.excludeFilters` worked great, now it's `AvoidInfix` with some listed exceptions

dwijnand replied:
"some" :D

helena commented on .scalafmt.conf:
```diff
@@ -0,0 +1,13 @@
+style = defaultWithAlign
+
+align                      = some
+align.tokens               = [off]
```
`align.tokens = [{code = "=>", owner = "Case"}]` [might be nice](https://scalameta.org/scalafmt/docs/configuration.html#aligntokens)

patriknw replied:
thanks, I was looking for that and thought it would be covered by `align = some`.
added in b042e06

helena commented:
Looks good, just a few suggestions to consider, or not.

helena reviewed (approved):
Added just a few things to consider first, otherwise LGTM

patriknw commented:
> enable the formatting for `MutiJvm` configuration

@2m seems like that is not enough, in the end I'd like it to be done via `test:compile`

patriknw commented:
I will not keep this updated for each and every change in master, so please review although it has conflicts. Before merging I will fix the conflicts ofc.

jrudolph commented:
> The performance seems ok since it does a good job at format incrementally.
> 
> a full `scalafmt` takes ~50 seconds

What does that mean in practice when switching branches? If the cache is lost in that case, adding 50 seconds to such a build is still a lot.

Disabling format-on-compile somewhat disrupts my workflow but since I'm using IntelliJ anyway, it might still be a net-positive to just integrate it properly into IntelliJ. Since the team mostly uses IntelliJ I guess it would be fine.

Also, on akka-http the automatic pr-formatter [now seems to work](https://github.com/akka/akka-http/pull/2425/commits/e1d35971e341255c401a25edf7f7a1255f0c8ec9), so even if the formatter isn't enabled automatically, we can create such a task for Akka as well and let it auto-format code and amend PRs.

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestKitUtils.scala:
```diff
@@ -18,25 +18,27 @@ import scala.concurrent.duration.Duration
-  final case class SpawnActor[T](name: String, behavior: Behavior[T], replyTo: ActorRef[ActorRef[T]], props: Props) extends TestKitCommand
-  final case class SpawnActorAnonymous[T](behavior: Behavior[T], replyTo: ActorRef[ActorRef[T]], props: Props) extends TestKitCommand
+  final case class SpawnActor[T](name: String, behavior: Behavior[T], replyTo: ActorRef[ActorRef[T]], props: Props)
+      extends TestKitCommand
```
What do we think about automatically adding line breaks like these?

helena replied:
I don't like it here and it ads 4 spaces, should be 2. 
This may be configurable with `verticalMultiline.arityThreshold`

patriknw replied:
I would prefer if it stayed on one line if within the 120 chars margin. Can't find a setting for it, do you know?

patriknw replied:
I think it's actually doing the right thing here, the line exceeds 120 and then that is a good place to break it.

I was worried that it always breaks before `extends`

dwijnand replied:
> I don't like it here and it ads 4 spaces, should be 2. Perhaps the default is 4.
> `continuationIndent.defnSite = 2`
> ?

Bodies are indented 2, so continuations being indented 4 is a nice visual guide that it's not the body.  IMHO

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -43,26 +43,30 @@ private[akka] object TestProbeImpl {
-        case other ⇒
-          queue.offerLast(other)
+    Behaviors
+      .receive[M] { (context, msg) =>
```
Why this?

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -43,26 +43,30 @@ private[akka] object TestProbeImpl {
-        case other ⇒
-          queue.offerLast(other)
+    Behaviors
+      .receive[M] { (context, msg) =>
```
Why this?

helena replied:
Let's not allow that format, though it's nice for other things like
```scala
  collection
    .func1(...)
    .func2(...)
    .func3
```

patriknw replied:
I agree, I hope it keeps the line breaks if they are there. I'll try
```
optIn.breakChainOnFirstMethodDot = false
```

patriknw replied:
didn't work can't find the setting for that, we would have to ask

jrudolph replied:
Agreed

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -118,18 +123,19 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
 
-    val ret = try f finally end = prevEnd
+    val ret = try f
+    finally end = prevEnd
```
Weird formatting here

helena replied:
Let's not do that. Something with `Term`?
https://github.com/scalameta/scalafmt/pull/1187

patriknw replied:
couldn't find a cure for this one

dwijnand replied:
Yeah, long-standing bug.

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -118,18 +123,19 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
     lastWasNoMessage = false
 
-    val maxDiff = max min rem
+    val maxDiff = max.min(rem)
```
I guess we just accept this style.

helena replied:
I think @patriknw pushes for that over spaces

patriknw replied:
I prefer `abc.map {` over `abc map {`
For some short operator style method names like `min` it can also be good, but one have to choose.

If we dislike this, we could run one sweep to get rid of infix and then turn off that formatting rule and manually add back infix where it makes sense.

patriknw replied:
maybe `rewrite.neverInfix.excludeFilters` could work

patriknw replied:
`rewrite.neverInfix.excludeFilters` worked great

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -170,8 +176,7 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
   def receiveMessage_internal(max: FiniteDuration): M =
-    receiveOne_internal(max).
-      getOrElse(assertFail(s"Timeout ($max) during receiveMessage while waiting for message."))
+    receiveOne_internal(max).getOrElse(assertFail(s"Timeout ($max) during receiveMessage while waiting for message."))
```
?

patriknw replied:
it's 119 chars, so fits on one line

(I would have expected it to preserve existing line breaks though)

patriknw replied:
now it's not touching this, not sure what I changed, perhaps the version

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/TestProbeImpl.scala:
```diff
@@ -244,51 +249,55 @@ private[akka] final class TestProbeImpl[M](name: String, system: ActorSystem[_])
-                s"seen messages ${seen.reverse}, hint: $hint", ex)
+        case Some(message) =>
+          val outcome = try fisher(message)
+          catch {
```
Weird indenting again. Can we have

```scala
val outcome =
  try fisher(message)
  catch {
  }
```

?

helena replied:
+1 that `try catch` line and indent change is bad

patriknw replied:
I agree, but how?

patriknw replied:
fixed this manually before formatting

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ManualTime.scala:
```diff
@@ -29,9 +29,12 @@ object ManualTime {
-        "make sure to include akka.actor.testkit.typed.javadsl.ManualTime.config() when setting up the test")
+      case sc: akka.testkit.ExplicitlyTriggeredScheduler => new ManualTime(sc)
+      case _ =>
+        throw new IllegalArgumentException(
```
Good one.

helena replied:
This
```suggestion
  case sc: akka.testkit.ExplicitlyTriggeredScheduler =>
    new ManualTime(sc)
  case _ =>
    throw new IllegalArgumentException(
```

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -180,12 +177,15 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
-  }
+  def stop[T](ref: ActorRef[T], max: FiniteDuration = timeout.duration): Unit =
+    try {
+      Await.result(internalSystem ? { x: ActorRef[ActorTestKitGuardian.Ack.type] =>
```
Not sure that's how I would break lines here? If it's multiple lines, then rather each argument on an extra line.

helena replied:
Depends on what follows it, e.g.
```scala
 def stop[T](ref: ActorRef[T], max: FiniteDuration = timeout.duration): Unit =
    try Await.result(internalSystem ? { x: ActorRef[ActorTestKitGuardian.Ack.type] =>
```
or

```scala
 def stop[T](ref: ActorRef[T], max: FiniteDuration = timeout.duration): Unit =
    try Await.result(internalSystem ? { 
      x: ActorRef[ActorTestKitGuardian.Ack.type] =>
```

jrudolph commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/Effects.scala:
```diff
@@ -4,7 +4,7 @@
 package akka.actor.testkit.typed.scaladsl
 
-import akka.actor.typed.{ ActorRef, Behavior, Props }
+import akka.actor.typed.{ActorRef, Behavior, Props}
```
Can we remove that setting? Seems like a useless change that just might get us more conflicts.

helena replied:
+1, the current formatting is nice, it ads that padding

patriknw replied:
fixed: `spaces.inImportCurlyBraces = true`

jrudolph reviewed (commented):
I looked through a few files. A few changes are good but it would be good if we could prevent other ones.

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -72,13 +75,11 @@ object CapturedLogEvent {
   private def toOption[A](optionVal: OptionVal[A]): Option[A] = optionVal match {
-    case OptionVal.Some(x) ⇒ Some(x)
-    case _                 ⇒ None
+    case OptionVal.Some(x) => Some(x)
```
👍

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/CapturedLogEvent.scala:
```diff
@@ -72,13 +75,11 @@ object CapturedLogEvent {
-  def apply(
-    logLevel: LogLevel,
-    message:  String): CapturedLogEvent = {
+  def apply(logLevel: LogLevel, message: String): CapturedLogEvent = {
```
better, for 2 args

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala:
```diff
@@ -25,18 +25,21 @@ import scala.concurrent.duration.FiniteDuration
 abstract class Effect private[akka] ()
 
 object Effect {
+
```
Yes, a space from type def before inner code is good. I see a lot w/out a space.

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala:
```diff
@@ -73,7 +79,8 @@ object Effect {
   object SpawnedAnonymous {
-    def apply[T](behavior: Behavior[T], props: Props = Props.empty): SpawnedAnonymous[T] = new SpawnedAnonymous(behavior, props, null)
+    def apply[T](behavior: Behavior[T], props: Props = Props.empty): SpawnedAnonymous[T] =
+      new SpawnedAnonymous(behavior, props, null)
```
So much better

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/Effect.scala:
```diff
@@ -167,6 +179,7 @@ object Effect {
    * The behavior set a new receive timeout, with `message` as timeout notification
    */
   final case class ReceiveTimeoutSet[T](d: FiniteDuration, message: T) extends Effect {
+
```
Line after is also so much better

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala:
```diff
@@ -179,10 +186,12 @@ private[akka] final class FunctionRef[-T](
         s"but [$child] is not a child of [$self]. Stopping other actors has to be expressed as " +
-        "an explicit stop message that the actor accepts.")
+        "an explicit stop message that the actor accepts."
+      )
```
Not a fan of the trailing new line for `)` can we configure that?

patriknw replied:
I agree, fixed: `danglingParentheses = false`

helena reviewed (changes requested):
Looking good. Added a few  areas to change the configuration on.

patriknw commented:
adjusted config some more: 355649e 
and updated to newer version: 49357b6

patriknw commented:
@2m my bad, your advice for formatting for `MutiJvm` worked, thanks

raboof commented:
> a full `scalafmt` takes ~50 seconds
> 
> If it turns out to be too slow we can disable format on compile, which is what is recommended in docs https://scalameta.org/scalafmt/docs/installation.html#format-on-compile.

It does look like switching to another branch makes the formatting run again... not sure if we want to accept that. OTOH perhaps it is OK to disable format-on-compile since scalafmt editor plugins are more ubiquitous?

I tried with sbt-scalafmt master and https://github.com/scalameta/sbt-scalafmt/pull/13 indeed seems to make a large difference. Filed https://github.com/scalameta/sbt-scalafmt/issues/19.

patriknw commented:
@raboof Let's use it without format on compile to avoid slowdown and that's also what is recommended by scalafmt docs. Format on save from IntelliJ instead.

I have also added formatting to the all* command aliases.

Defined version to 2.0.0-RC5

and rebased

**Ready for final approval or rejection**

BTW, I will squash the commits before merging, but keeping the "format source" separate and also see if I can use another bot author for that commit (I don't want to take credit for that big changes when it's only by tool).

helena commented:
@patriknw what about anyone not using IntelliJ - if you're making the assumption.

dwijnand commented:
Setting the version to 2.0.0-RC5 and then using sbt-scalafmt 1.6.0-RC4 is probably not great.  Note that there is no version of sbt-scalafmt for Scalafmt 2 yet (tracked as https://github.com/scalameta/sbt-scalafmt/issues/15).

For Play, I opted for using Scalafmt 2 by configuring CI to use the CLI launcher.

jrudolph commented on build.sbt:
```diff
@@ -524,37 +524,42 @@ def akkaModule(name: String): Project =
-addCommandAlias("allDistributedData", commandValue(distributedData))
-addCommandAlias("allClusterSharding", commandValue(clusterSharding))
-addCommandAlias("allClusterTools", commandValue(clusterTools))
+addCommandAlias("allRemote", commandValue(remote, Some(remoteTests), multiJvmTests = true))
```
That's nice!

jrudolph commented:
@patriknw can you sum up how the new workflow will be? (It's hard to say because build formatting changes makes it hard to review any actual build changes).

 * How does the developer use it? Usually, with IntelliJ IIUC?
 * How do we enforce correct formatting in PRs? Does it fail fast?
 * Do we need to update any contribution docs, so that contributors will by default do the right thing?

patriknw commented:
@dwijnand Thanks for clarifying that. I thought it was supported because the tickets mentioned that a release was already done. I was some other tickets mentioning scalafmt-dynamic and guessed it was using that already since I couldn't find any reference to a scalafmt version anywhere.

> probably not great

What does that mean? It doesn't fail so is it just that it's not using the specified version?

What version is supported by the IntelliJ plugin?

dwijnand commented:
> What does that mean? It doesn't fail so is it just that it's not using the specified version?
> 
> What version is supported by the IntelliJ plugin?

Only as of Scalafmt 2.0.0-RC5 does `scalafmt-dynamic` use the version defined in `.scalafmt.conf` to dispatch to the right version of Scalafmt.  https://github.com/scalameta/scalafmt/commit/fd48c4b0ea33608923dcbf78ec4bfafdb06699bd

IntelliJ currently supports latest 1.5 (1.5.1) and latest 1.6 RC (1.6.0-RC4).  I think development on the 1.x line has stopped and the setup around the 2.x line is coming.  For instance, https://github.com/scalameta/scalafmt/commit/fd840e7d8e4d1371e731447e3baf710a0343a7b5 where work was done for the IntelliJ integration.

So you can either:
* hold this PR, wait for 2.x to be fully ready, then merge
* merge this PR, as 2.x becomes more ready move to a preferred workflow
* switch to 1.6, reformat, then when 2.x is ready upgrade (and probably reformat some differently-formatting sources)

2m commented:
The sbt plugin with enhancements has been released. Note that org for the artifacts has changed as well:

```scala
addSbtPlugin("org.scalameta" % "sbt-scalafmt" % "2.0.0-RC5")
```

dwijnand commented:
Yeah, I just resolved it locally.  I'm switch play-grpc to it.  The check tasks are `scalafmtCheckAll` and `scalafmtSbtCheck`.

dwijnand commented:
The IntelliJ plugin has also been updated: https://plugins.jetbrains.com/plugin/8236-scalafmt/update/59101

So the concerns I had shared here to forewarn about adopting Scalafmt 2 have all been resolved.  🕺

Commit:
format source with scalafmt

patriknw commented:
I have updated to sbt-scalafmt 2.0.0-RC5 and rebased.
I tried the IntelliJ nigtly plugin and it seems to work.

@jrudolph The workflow is to use the IntelliJ plugin and configure it to format on save. Then nothing else should be needed. I added note in CONTRIBUTING about the plugin.

If not using the plugin the sources can be formatted from `sbt` with `scalafmtAll`, typically on individual projects that you have changed. I added note in CONTRIBUTING about this.

I also added `scalafmtAll` to the `all*` command aliases.

We currently have a `git diff` check in jenkins pr validator job. That can be replaced by `clean scalafmtCheckAll test:compile`, which will fail the build earlier than before. We probably have to keep the `git diff` to detect java formatting issues.

patriknw committed:
fix wrong headers

patriknw commented:
RoutersSpec

patriknw commented:
PLS BUILD

johanandren commented:
@patriknw that RoutersSpec has been failing all weekend I have fixed it in a PR today so maybe just go ahead and merge anyways?

patriknw commented:
I will not merge this (or anything) without green PR validation.

johanandren commented:
In that case, could be good to merge #26497 which fixes the `RoutersSpec`
